File: MomentsAndDeviation/Median.cpp
/*

Vázquez Choreño Luis Ernesto
Randomized Algorithms
Application: Median

g++ -std=c++11 Median.cpp  && ./a.out

*/

#include <bits/stdc++.h>
using namespace std;
const int inf_int = 10000;


int get_rng(int value){ return (rand()%value ) + 1;}

/// MEDIAN ALGORITHM
vector<int> generate_random_array(int n){
  vector<int> v(n);
  for(int i=0;i<n;i++){
    v[i] = get_rng(inf_int);
  }
  return v;
}


void process_agorithm_randomized(vector<int> &S){
  bool found = false;
  int n = S.size();
  while(found == false){
    int size_r = ceil(pow(n,0.75));
    vector<int> R(size_r);
    // first step
    for(int i=0;i<R.size();i++){
      R[i] = S[get_rng(n) - 1];
    }

    // second step
    sort(R.begin(),R.end());

    // third step
    int d = R[floor(size_r/2.00 - pow(n,0.5))];

    // fourth step
    int u = R[ ceil(size_r/2.00 + pow(n,0.5))];


    // fifth step
    vector<int> C;
    int ld = 0,lu = 0;
    for(int i=0;i<n;i++){
      if(d <= S[i] and S[i]<=u){
        C.push_back(S[i]);
      }else if(S[i] < d){
        ld++;
      }else{
        lu++;
      }
    }


    // sixth step
    if(ld > (n>>1) || lu > (n>>1))
      continue;

    // seventh step
    if(C.size() > 4*size_r)
      continue;
    sort(C.begin(),C.end());

    // eight step
    int median_index = (n>>1) - ld ;
    cout << "Median computed: " << C[median_index] << "\n";
    found = true;
  }
}




/// R-SELECT

int partition(vector<int> &S,int l,int r){
  int value = S[r];
  int i = l;
  for(int j = l; j < r; j++){
    if(S[j] <= value){
      swap(S[i],S[j]);
      i++;
    }
  }
  swap(S[i],S[r]);
  return i;
}


int r_select(vector<int> &S,int l,int r,int k){
  int len = r - l + 1;
  int pivot = get_rng(len) - 1 + l;
  swap(S[pivot],S[r]);
  int pos =  partition(S,l,r);
  if(pos - l == k)
    return S[pos];
  if(pos - l > k)
    return r_select(S,l,pos-1,k);
  return r_select(S,pos+1,r,k-pos+l-1);
}


void r_select(vector<int> &S){
  int len = S.size();
  int median = r_select(S,0,len-1,(len)>>1);
  cout << "Median computed: " << median << "\n";
}









int main(){

  srand(time(NULL));
  int n = 333;
  vector<int> random_array = generate_random_array(n);
  vector<int> copy_array(random_array);
  sort(copy_array.begin(),copy_array.end());
  cout << "Real median: " << copy_array[n>>1] << "\n";


  process_agorithm_randomized(random_array);
  r_select(random_array);

  return 0;
}

File: MomentsAndDeviation/Median.py
'''
Vazquez Choreno Luis Ernesto
Randomized Algorithms
Application: Median

python3 Median.py
'''

import numpy as np
import math
import matplotlib.pyplot as plt


limit = 10000
number_operations = 0
number_operations2 = 0


def get_rng(value):
    return np.random.randint(low=1, high=value+1)


### MEDIAN ALGORITHM ###
def generate_random_array(n):
    S = []
    for i in range(n):
        S.append(get_rng(limit))
    return S



def process_agorithm_randomized(S):
    found = False
    global number_operations2
    number_operations2 = 0
    n = len(S)
    while found == False:
        size_r = int(np.ceil(pow(n,0.75)))
        R = []

        # first step
        number_operations2 += int(size_r)
        for i in range(size_r):
            R.append(S[get_rng(n) - 1])

        #second step
        number_operations2 += int(size_r * math.log2(size_r))
        R.sort()


        #third step
        d = R[int(np.floor(size_r/2.00 - pow(n,0.5)))]

        #fourth step
        u = R[int( np.ceil(size_r/2.00 + pow(n,0.5)))]

        #fifth step
        number_operations2 += len(S)
        C = []
        ld = 0
        lu = 0
        for si in S:
            if d <= si and si <= u:
                C.append(si)
            elif si < d:
                ld+=1
            else:
                lu+=1
        # sixth step
        if ld > int(n/2) or lu > int(n/2):
            continue

        #seventh step
        if len(C) > 4*size_r:
            continue
        C.sort()
        number_operations2 += int(len(C)*math.log2(len(C)))

        ##eight step
        median_index = int(n/2.00) - ld
        print("Median computed: " + str(C[median_index]))
        found = True
    return number_operations2






### R-SELECT ###
def partition(S,l,r):
    value = S[r]
    i = l
    for j in range(l,r):
        if S[j] <= value:
            temp = S[i]
            S[i] = S[j]
            S[j] = temp
            i+=1
    temp = S[i]
    S[i] = S[r]
    S[r] = temp
    return i



def rselect(S,l,r,k):
    L = r - l + 1
    pivot = get_rng(L) - 1 + l
    temp = S[pivot]
    S[pivot] = S[r]
    S[r] = temp
    global number_operations
    number_operations += L
    pos = partition(S,l,r)
    if pos - l == k:
        return S[pos]
    if pos - l > k:
        return rselect(S,l,pos-1,k)
    return rselect(S,pos+1,r,k-pos+l-1)

def r_select(S):
    L = len(S)
    global number_operations
    number_operations = 0
    median = rselect(S,0,L-1,int(np.floor(L/2.00)))
    print("Median computed: " + str(median))
    return number_operations







def main():
    n = 100
    S = generate_random_array(n)
    copyS = S.copy()
    copyS.sort()
    print("Real median: " + str(copyS[int(np.floor(n/2.00))]))
    r_select(S)
    process_agorithm_randomized(S)




    '''
    x = []
    y = []
    y1 = []
    for i in range(1000,10000,1000):
        n = i
        x.append(i)
        yi = 0
        y1i = 0
        for j in range(0,10):
            S = generate_random_array(n)
            yi += r_select(S)
            y1i += process_agorithm_randomized(S)
        y.append(yi/10.00)
        y1.append(y1i/10.00)

    plt.plot(x,y,label = 'R-SELECT')
    plt.plot(x,y1,label = 'MEDIAN ALGORITHM')
    plt.plot(x,x, label = 'LINEAR TIME')
    plt.legend()
    plt.show()
    '''
main()
